### PR TITLE
Update/SOF-7508 update: fix passivation with coordinatation method

### PR DESCRIPTION
**Pull request by VsevolodX**

### CODE REVIEW

File: src/py/mat3ra/made/tools/build/passivation/builders.py [hunk 1]
```diff
@@ -130,70 +129,231 @@ class CoordinationBasedPassivationBuilderParameters(SurfacePassivationBuilderPar
...
+                    )
+            else:
+                # Fallback to nearest neighbors average vector method if no reconstructed bonds found
+                vectors = nearest_neighbors_coordinates[idx]
+                bond_vector = -np.mean(vectors, axis=0)
+                bond_vector = bond_vector / np.linalg.norm(bond_vector) * configuration.bond_length
+                passivant_bond_vector_crystal = material.basis.cell.convert_point_to_crystal(bond_vector)
+                passivant_coordinates.append(
+                    material.basis.coordinates.get_element_value_by_index(idx) + passivant_bond_vector_crystal
+                )
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/passivation/builders.py:356]: We need tests for all the above functionality

File: src/py/mat3ra/made/tools/build/passivation/coordination.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                    self.are_bonds_templates_similar(template, existing, self.angle_tolerance)
+                    for existing in unique_templates
+                ):
+                    unique_templates.append(template)
+
+            element_templates[element] = unique_templates
+
+        return element_templates
+
+    def reconstruct_missing_bonds(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/passivation/coordination.py:136]: isolate reconstruct_missing_bonds_for_atom and have `reconstruct_missing_bonds` for the whole material

File: src/py/mat3ra/made/tools/build/passivation/coordination.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
+from typing import Dict, List
+
+import numpy as np
+from pydantic import BaseModel, Field
+
+from mat3ra.made.material import Material
+from mat3ra.made.tools.analyze import get_nearest_neighbors_vectors
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/passivation/coordination.py:7]: Should be part of this file

File: src/py/mat3ra/made/tools/build/passivation/coordination.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
+from typing import Dict, List
+
+import numpy as np
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/passivation/coordination.py:3]: Let's convert to namespace

File: src/py/mat3ra/made/tools/build/passivation/coordination.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        element_templates = {}
+
+        for element in set(atom_elements):
+            element_indices = [i for i, e in enumerate(atom_elements) if e == element]
+            element_vector_lists = [np.array(atom_vectors[i]) for i in element_indices]
+
+            if not element_vector_lists:
+                continue
+
+            max_coord = max(len(vectors) for vectors in element_vector_lists)
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/passivation/coordination.py:121]: max_coordination_number

Same below

File: tests/py/unit/test_tools_build_passivation.py [hunk 6]
```diff
@@ -13,3 +19,21 @@ def test_passivate_surface():
...
+
+def test_get_unique_coordination_numbers():
+    config = PassivationConfiguration(
+        slab=Material(GRAPHENE_ZIGZAG_NANORIBBON), passivant="H", bond_length=1.48, surface="both"
+    )
+    unique_coordination_numbers = get_unique_coordination_numbers(config, cutoff=3.0)
+    assert unique_coordination_numbers == [2, 3]
+
+
+def test_passivate_edge():
```
Comment (timurbazhirov) [inline tests/py/unit/test_tools_build_passivation.py:32]: test_passivate_coordination_based

File: src/py/mat3ra/made/tools/build/passivation/coordination.py (hunk 2 above)
Comment (VsevolodX) [reply src/py/mat3ra/made/tools/build/passivation/coordination.py]: Ok. And this is where we'll pass bond templates distinguished by element

File: src/py/mat3ra/made/tools/analyze/__init__.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+    @property
+    def volume(self):
+        return self.material.lattice.volume
+
+    @property
+    def atomic_density(self):
+        return len(self.material.coordinates_array) / self.volume()
+
+    def pairwise_distances(self):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/__init__.py:17]: Needs to be a property too

File: src/py/mat3ra/made/tools/analyze/__init__.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+from scipy.spatial.distance import pdist
+
+
+class BaseMaterialAnalyzer:
+    def __init__(self, material: Material):
+        self.material = material
+
+    @property
+    def volume(self):
+        return self.material.lattice.volume
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/__init__.py:11]: needs to come from basis.cell

File: src/py/mat3ra/made/tools/analyze/coordination.py [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+        super().__init__(**data)
+        self.nearest_neighbor_vectors = self.get_nearest_neighbors_vectors()
+        self.crystal_sites = CrystalSiteList(nearest_neighbor_vectors=self.nearest_neighbor_vectors)
+
+    @property
+    def coordinates_as_kdtree(self):
+        return cKDTree(self.basis.coordinates.values)
+
+    @decorator_handle_periodic_boundary_conditions(0.25)
+    def get_nearest_neighbors_vectors(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/coordination.py:90]: Needs to be removed

File: src/py/mat3ra/made/tools/analyze/coordination.py [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+class CrystalSiteList(ArrayWithIds):
+    values: List[CrystalSite]
+
+
+class MaterialWithCrystalSites(Material):
+    crystal_sites: CrystalSiteList = CrystalSiteList()
+
+    def __init__(self, **data):
+        super().__init__(**data)
+        self.nearest_neighbor_vectors = self.get_nearest_neighbors_vectors()
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/coordination.py:82]: get_nearest_neighbors_vectors(self, cutoff <from data>)

File: src/py/mat3ra/made/tools/analyze/coordination.py [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+        atom_elements (List[str]): List of chemical elements for each atom.
+        angle_tolerance (float): Tolerance for comparing angles between bond vectors.
+
+    Returns:
+        Dict[str, List[np.ndarray]]: Dictionary mapping element types to unique bond templates.
+    """
+    element_templates = {}
+
+    for element in set(material.basis.elements.values):
+        element_templates[element] = find_bond_directions_for_element(material, element, angle_tolerance)
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/coordination.py:298]: We need to define a structure that will hold this - I'm suggesting a flat list with {element: "Si", bond_direction: BondDirection()} as members

File: src/py/mat3ra/made/tools/analyze/coordination.py [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+            dot_matrix = np.dot(template, existing_vectors.T)
+            cosine_matrix = dot_matrix / (
+                np.linalg.norm(template, axis=1)[:, None] * np.linalg.norm(existing_vectors, axis=1)
+            )
+            angles_matrix = np.arccos(np.clip(cosine_matrix, -1.0, 1.0))
+
+            matches = np.any(angles_matrix < angle_tolerance, axis=1)
+            match_count = np.sum(matches)
+
+        missing = template[~matches] if existing_vectors.size != 0 else template
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/coordination.py:382]: This is duplicate, can be just equality check now

File: src/py/mat3ra/made/tools/analyze/coordination.py [hunk 13]
```diff
@@ -0,0 +1,499 @@
...
+
+    unique_templates: List[np.ndarray] = []
+    for template in max_coordination_number_vectors:
+        if all(template != existing for existing in unique_templates):
+            unique_templates.append(template)
+
+    return unique_templates
+
+
+def reconstruct_missing_bonds_for_element(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/coordination.py:337]: Could be called `find_missing_directions` and go to BondDirections class

File: src/py/mat3ra/made/tools/analyze/coordination.py [hunk 14]
```diff
@@ -0,0 +1,499 @@
...
+            len(best_missing),
+            max_bonds_to_passivate,
+            max_coordination_number - len(existing_vectors),
+        )
+        return best_missing[:num_bonds_to_add].tolist()
+
+    return []
+
+
+def reconstruct_missing_bonds(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/coordination.py:399]: find_missing_bonds_for_all_sites -> goto MaterialWithCrystalSites

File: src/py/mat3ra/made/tools/analyze/coordination.py (hunk 10 above)
Comment (VsevolodX) [reply src/py/mat3ra/made/tools/analyze/coordination.py:82]: Maybe this? self is already passed
```        
self.nearest_neighbor_vectors = self.get_neighbors_vectors_for_all_sites(cutoff=3.0)
```

File: src/py/mat3ra/made/tools/analyze/coordination.py [hunk 15]
```diff
@@ -0,0 +1,467 @@
...
+            if len(matches) == 0:
+                return False
+            unmatched.remove(matches.tolist()[0])
+
+        return True
+
+    @staticmethod
+    def find_missing_directions(
+        vectors: List[np.ndarray],
+        element: str,
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/coordination.py:62]: This should only deal with a single template and should not have elements

File: src/py/mat3ra/made/tools/analyze/coordination.py [hunk 16]
```diff
@@ -0,0 +1,467 @@
...
+            Dict[int, int]: A dictionary mapping atom indices to their coordination numbers.
+        """
+        nearest_neighbors = self.get_nearest_neighbors_for_all_sites(cutoff)
+        coordination_numbers = ArrayWithIds(values=nearest_neighbors)
+        return coordination_numbers
+
+    def find_missing_bonds_for_all_sites(self, templates: Dict[str, List[np.ndarray]]) -> Dict[int, List[List[float]]]:
+        missing_bonds = {}
+        for idx, (vectors) in enumerate(self.nearest_neighbor_vectors):
+            reconstructed_bonds = BondDirections.find_missing_directions(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/coordination.py:250]: The per-element logic should go here

File: src/py/mat3ra/made/tools/analyze/rdf.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
+import numpy as np
+from mat3ra.made.material import Material
+from pydantic import BaseModel
+
+from . import BaseMaterialAnalyzer
+
+
+class RadialDistributionFunction(BaseModel):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/rdf.py:8]: We should have a test for this

File: src/py/mat3ra/made/cell.py [hunk 18]
```diff
@@ -51,3 +51,7 @@ def convert_point_to_crystal(self, point: List[float]) -> List[float]:
     def scale_by_matrix(self, matrix: List[List[float]]):
         np_vector = np.array(self.vectors_as_array)
         self.vector1, self.vector2, self.vector3 = np.dot(np.array(matrix), np_vector).tolist()
+
+    @property
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/cell.py:55]: We don't need this `__init.py__` file in `src`

File: src/py/mat3ra/made/tools/analyze/material.py [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+    def get_neighbors_vectors_for_all_sites(
+        self, cutoff: float = 3.0, max_number_of_neighbors: Optional[int] = None
+    ) -> ArrayWithIds:
+        nearest_neighbors = ArrayWithIds()
+        for site_index in range(len(self.basis.coordinates.values)):
+            vectors = self.get_neighbors_vectors_for_site(site_index, cutoff, max_number_of_neighbors)
+            nearest_neighbors.add_item(vectors, site_index)
+        return nearest_neighbors
+
+    def get_neighbors_vectors_for_site(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/material.py:44]: should be above `get_neighbors_vectors_for_all_sites`

File: src/py/mat3ra/made/tools/analyze/material.py [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+            cutoff (float): The maximum cutoff radius for identifying neighbors.
+            max_number_of_neighbors (int): The max number of neighbors possible.
+        """
+        nearest_neighbors = ArrayWithIds()
+        for site_index in self.basis.coordinates.ids:
+            neighbors, distances = self.get_neighbors_for_site(site_index, cutoff, max_number_of_neighbors)
+            nearest_neighbors.add_item(neighbors, site_index)
+        return nearest_neighbors
+
+    def get_neighbors_for_site(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/material.py:68]: Same as the above

File: src/py/mat3ra/made/tools/analyze/material.py [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+
+    def get_neighbors_vectors_for_site(
+        self, site_index: int, cutoff: float = 3.0, max_number_of_neighbors: Optional[int] = None
+    ) -> List[np.ndarray]:
+        coordinates = self.basis.coordinates.values
+        neighbors, distances = self.get_neighbors_for_site(site_index, cutoff, max_number_of_neighbors)
+        vectors = [np.array(coordinates[neighbor]) - np.array(coordinates[site_index]) for neighbor in neighbors]
+        return vectors
+
+    def get_nearest_neighbors_for_all_sites(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze/material.py:52]: Shouldn't this fn use the decorator too?

File: tests/py/unit/fixtures.py [hunk 22]
```diff
@@ -492,6 +492,99 @@
     "isUpdated": True,
 }
 
+GRAPHENE_ZIGZAG_NANORIBBON_PASSIVATED = {
```
Comment (timurbazhirov) [inline tests/py/unit/fixtures.py:495]: Nit: we can put this to Standata and import here - will be easier to visualize, etc.

File: src/py/mat3ra/made/tools/analyze/rdf.py (hunk 17 above)
Comment (timurbazhirov) [reply src/py/mat3ra/made/tools/analyze/rdf.py:8]: Still no test

**Review by timurbazhirov — APPROVED** [review]

**Pull request merged**